File: examples/nodejs/secret-manager-factory.js
/**
 * Secret Manager Factory — Abstracting Away Provider Details
 *
 * WHY A FACTORY PATTERN?
 * ───────────────────────
 * 1. AVOID VENDOR LOCK-IN: Your application code shouldn't care whether secrets
 *    come from AWS, Azure, or Vault. If you switch cloud providers, you only
 *    change one environment variable (SECRET_PROVIDER), not your entire codebase.
 *
 * 2. CONSISTENT INTERFACE: All providers implement the same methods, so
 *    calling code is identical regardless of backend.
 *
 * 3. EASY TESTING: Swap in a mock provider for unit tests without touching
 *    production code or needing real credentials.
 *
 * 4. MULTI-CLOUD: In complex organizations, different teams may use different
 *    providers. The factory lets each team configure their own while sharing
 *    the same application code.
 */

const winston = require("winston");

// ── Logger Configuration ──────────────────────────────────────────────────
const logger = winston.createLogger({
  level: process.env.LOG_LEVEL || "info",
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.errors({ stack: true }),
    winston.format.json()
  ),
  transports: [new winston.transports.Console()],
});

// ── Base SecretManager Interface ──────────────────────────────────────────
// This defines the contract that ALL providers must fulfill.
// In TypeScript this would be an interface; in JavaScript we document it clearly.

/**
 * @abstract
 * Base class defining the secret manager interface.
 * All provider implementations must extend this class.
 */
class SecretManager {
  /**
   * Fetches a generic secret by name.
   * @param {string} name - Secret identifier (name, path, or ARN).
   * @returns {Promise<object|string>} The secret value.
   */
  async getSecret(name) {
    throw new Error("getSecret() must be implemented by subclass");
  }

  /**
   * Fetches database credentials.
   * @returns {Promise<object>} Database connection parameters.
   */
  async getDatabaseCredentials() {
    throw new Error("getDatabaseCredentials() must be implemented by subclass");
  }

  /**
   * Fetches JWT signing configuration.
   * @returns {Promise<object>} JWT configuration with algorithm and secret.
   */
  async getJwtConfig() {
    throw new Error("getJwtConfig() must be implemented by subclass");
  }

  /**
   * Returns the provider name for logging/debugging.
   * @returns {string}
   */
  getProviderName() {
    throw new Error("getProviderName() must be implemented by subclass");
  }
}

// ── AWS Secrets Manager Implementation ────────────────────────────────────
class AWSSecretManager extends SecretManager {
  constructor() {
    super();
    // Lazy import to avoid requiring @aws-sdk/client-secrets-manager
    // when using a different provider
    this._aws = null;
  }

  _getModule() {
    if (!this._aws) {
      this._aws = require("./aws-secrets");
    }
    return this._aws;
  }

  async getSecret(name) {
    return this._getModule().getSecret(name);
  }

  async getDatabaseCredentials() {
    return this._getModule().getDatabaseCredentials(
      process.env.AWS_DB_SECRET_NAME || "prod/database/credentials"
    );
  }

  async getJwtConfig() {
    return this._getModule().getJwtConfig(
      process.env.AWS_JWT_SECRET_NAME || "prod/jwt/config"
    );
  }

  getProviderName() {
    return "AWS Secrets Manager";
  }
}

// ── HashiCorp Vault Implementation ────────────────────────────────────────
class VaultSecretManager extends SecretManager {
  constructor() {
    super();
    this._vault = null;
  }

  _getModule() {
    if (!this._vault) {
      this._vault = require("./vault-secrets");
    }
    return this._vault;
  }

  async getSecret(name) {
    return this._getModule().getSecret(name);
  }

  async getDatabaseCredentials() {
    return this._getModule().getDatabaseCredentials(
      process.env.VAULT_DB_CREDS_PATH || "database/creds/app-readonly"
    );
  }

  async getJwtConfig() {
    return this._getModule().getJwtConfig(
      process.env.VAULT_JWT_PATH || "secret/data/app/jwt"
    );
  }

  getProviderName() {
    return "HashiCorp Vault";
  }
}

// ── Azure Key Vault Implementation ────────────────────────────────────────
class AzureSecretManager extends SecretManager {
  constructor() {
    super();
    this._azure = null;
  }

  _getModule() {
    if (!this._azure) {
      this._azure = require("./azure-secrets");
    }
    return this._azure;
  }

  /**
   * Azure Key Vault returns raw strings, so we wrap them in an object
   * to match the interface expected by the rest of the application.
   */
  async getSecret(name) {
    const value = await this._getModule().getSecret(name);
    return { value, name };
  }

  async getDatabaseCredentials() {
    // Azure convention: connection strings are single secret values
    const connStr = await this._getModule().getDatabaseConnectionString(
      process.env.AZURE_DB_SECRET_NAME || "DatabaseConnectionString"
    );
    return { connectionString: connStr };
  }

  async getJwtConfig() {
    const secretKey = await this._getModule().getJwtSigningKey(
      process.env.AZURE_JWT_SECRET_NAME || "JwtSigningKey"
    );
    return {
      algorithm: process.env.JWT_ALGORITHM || "HS256",
      secretKey,
      issuer: process.env.JWT_ISSUER || "secrets-management-app",
      expiresIn: process.env.JWT_EXPIRES_IN || "15m",
    };
  }

  getProviderName() {
    return "Azure Key Vault";
  }
}

// ── Factory Function ──────────────────────────────────────────────────────
// Supported provider values for the SECRET_PROVIDER environment variable:
//   "aws"     — AWS Secrets Manager
//   "vault"   — HashiCorp Vault
//   "azure"   — Azure Key Vault

/**
 * Creates a SecretManager instance based on the SECRET_PROVIDER environment variable.
 *
 * USAGE:
 *   SECRET_PROVIDER=aws node server.js
 *   SECRET_PROVIDER=vault node server.js
 *   SECRET_PROVIDER=azure node server.js
 *
 * The factory pattern ensures the rest of your application never imports
 * provider-specific modules directly — it only depends on the base interface.
 *
 * @param {string} [provider] - Override provider (defaults to SECRET_PROVIDER env var).
 * @returns {SecretManager} A configured secret manager instance.
 */
function createSecretManager(provider) {
  const selected = (provider || process.env.SECRET_PROVIDER || "").toLowerCase().trim();

  if (!selected) {
    throw new Error(
      "SECRET_PROVIDER environment variable is required.\n" +
      "Supported values: 'aws', 'vault', 'azure'\n" +
      "Example: SECRET_PROVIDER=aws node server.js"
    );
  }

  let manager;

  switch (selected) {
    case "aws":
      manager = new AWSSecretManager();
      break;
    case "vault":
      manager = new VaultSecretManager();
      break;
    case "azure":
      manager = new AzureSecretManager();
      break;
    default:
      throw new Error(
        `Unsupported SECRET_PROVIDER '${selected}'. ` +
        `Supported values: 'aws', 'vault', 'azure'`
      );
  }

  logger.info("Secret manager initialized via factory", {
    provider: selected,
    providerName: manager.getProviderName(),
  });

  return manager;
}

// Export both the factory and individual classes for direct use if needed
module.exports = {
  createSecretManager,
  SecretManager,
  AWSSecretManager,
  VaultSecretManager,
  AzureSecretManager,
};

File: examples/python/secret_manager_factory.py
"""
Secret Manager Factory
=======================
This module implements the *Factory pattern* to decouple application code
from any specific secrets provider (AWS, Vault, Azure).

Why this matters for security and maintainability
-------------------------------------------------
* **Zero vendor lock-in:** Switching from AWS to Vault (or vice versa)
  requires changing ONE environment variable (SECRET_PROVIDER) and
  ensuring the right SDK is installed — no application code changes.
* **Single responsibility:** Every provider implements the same interface.
  Application modules (auth, database, etc.) depend only on the abstract
  interface, never on concrete provider classes.
* **Testability:** In tests you can supply a mock or in-memory provider
  that implements `SecretManager` without spinning up real infrastructure.
* **Auditable:** The abstract base class documents exactly which secrets
  the application expects, making security reviews straightforward.
"""

import logging
import os
from abc import ABC, abstractmethod
from typing import Any

# ---------------------------------------------------------------------------
# Logging — same "no secrets in logs" policy as provider modules.
# ---------------------------------------------------------------------------
logger = logging.getLogger("secret_factory")
logger.setLevel(logging.INFO)


class SecretManager(ABC):
    """Abstract interface that every secrets provider must implement.

    This is the *contract* between the application and its secret store.
    Only methods that the application actually needs are included, keeping
    the surface area small and auditable.
    """

    @abstractmethod
    def get_secret(self, secret_name: str) -> dict[str, Any]:
        """Fetch a generic secret by name and return its data dict."""
        ...

    @abstractmethod
    def get_database_credentials(self) -> dict[str, Any]:
        """Fetch structured database connection parameters."""
        ...

    @abstractmethod
    def get_jwt_config(self) -> dict[str, Any]:
        """Fetch JWT signing configuration."""
        ...

    @abstractmethod
    def get_openai_config(self) -> dict[str, Any]:
        """Fetch OpenAI API configuration."""
        ...


# ---------------------------------------------------------------------------
# Provider-specific imports — we use lazy imports inside factory functions
# so that code depending only on the ABC (e.g. tests) does not require
# all three SDKs to be installed.
# ---------------------------------------------------------------------------

def _create_aws_manager() -> SecretManager:
    """Build an AWSSecretsManager instance.

    All configuration (region, secret names) comes from environment
    variables — see aws_secrets.py for details.
    """
    # Lazy import so Azure/Vault SDKs are not required if unused.
    from aws_secrets import AWSSecretsManager

    logger.info({"event": "creating_provider", "provider": "aws"})
    return AWSSecretsManager()


def _create_vault_manager() -> SecretManager:
    """Build a VaultSecretsManager instance.

    Requires VAULT_ADDR and VAULT_TOKEN environment variables.
    """
    from vault_secrets import VaultSecretsManager

    logger.info({"event": "creating_provider", "provider": "vault"})
    return VaultSecretsManager()


def _create_azure_manager() -> SecretManager:
    """Build an AzureKeyVaultManager instance.

    Requires AZURE_VAULT_URL.  Auth via DefaultAzureCredential.
    """
    from azure_secrets import AzureKeyVaultManager

    logger.info({"event": "creating_provider", "provider": "azure"})
    return AzureKeyVaultManager()


# ---------------------------------------------------------------------------
# Factory registry
# ---------------------------------------------------------------------------
_PROVIDERS: dict[str, type] = {
    "aws": None,   # populated lazily
    "vault": None,
    "azure": None,
}

# We store callables rather than classes to keep imports lazy.
_PROVIDER_FACTORIES: dict[str, Any] = {
    "aws": _create_aws_manager,
    "vault": _create_vault_manager,
    "azure": _create_azure_manager,
}


def get_secret_manager(provider: str | None = None) -> SecretManager:
    """Factory method: return a SecretManager for the given provider.

    The provider is determined by (in priority order):
      1. The ``provider`` argument passed to this function.
      2. The ``SECRET_PROVIDER`` environment variable.
      3. Raises ``ValueError`` if neither is set.

    Valid provider values: "aws", "vault", "azure".

    SECURITY: We log which provider was selected but never any
    configuration values (URLs, tokens, etc.) that could leak.
    """
    resolved = provider or os.getenv("SECRET_PROVIDER")

    if not resolved:
        raise ValueError(
            "No secret provider specified. Set the SECRET_PROVIDER "
            "environment variable to 'aws', 'vault', or 'azure'."
        )

    resolved = resolved.lower().strip()

    factory = _PROVIDER_FACTORIES.get(resolved)
    if factory is None:
        raise ValueError(
            f"Unknown secret provider '{resolved}'. "
            f"Supported: {list(_PROVIDER_FACTORIES.keys())}"
        )

    logger.info({"event": "provider_selected", "provider": resolved})
    return factory()


# ---------------------------------------------------------------------------
# Convenience: a module-level singleton (created on first access)
# ---------------------------------------------------------------------------
_manager_instance: SecretManager | None = None


def get_manager() -> SecretManager:
    """Return a cached SecretManager singleton.

    This avoids creating multiple clients (and thus multiple network
    connections) when different modules in the same process all need
    secrets.  The singleton is created on first call.
    """
    global _manager_instance
    if _manager_instance is None:
        _manager_instance = get_secret_manager()
    return _manager_instance


# ---------------------------------------------------------------------------
# Demonstration
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO, format="%(message)s")

    # Simulate selecting a provider via env var
    os.environ.setdefault("SECRET_PROVIDER", "aws")

    try:
        mgr = get_manager()
        print(f"Provider type: {type(mgr).__name__}")
        print("Factory pattern working — application code is provider-agnostic.")
    except Exception as exc:
        print(f"Expected error in local dev (no cloud creds): {exc}")